Pal/lib/list.h

```diff
@@ -232,46 +234,59 @@
         listp_del(NODE, HEAD, FIELD);           \
         INIT_LIST_HEAD(NODE, FIELD);            \
     } while(0)
-            
+
 /* Keep vestigial TYPE and FIELD parameters to minimize disruption
  * when switching from Linux list implementation */
 #define listp_first_entry(LISTP, TYPE, FIELD) ((LISTP)->first)
 
 /* New API: return last entry in list */
 #define listp_last_entry(LISTP, TYPE, FIELD) ((LISTP)->first->FIELD.prev)
 
+/* New API: return next entry in list */
+#define listp_next_entry(NODE, LISTP, FIELD)                            \
+        ((NODE) == (LISTP)->first->FIELD.prev ? NULL : (NODE)->FIELD.next)
+
+/* New API: return previous entry in list */
+#define listp_prev_entry(NODE, LISTP, FIELD)                            \
+        ((NODE) == (LISTP)->first ? NULL : (NODE)->FIELD.prev)
+
 /* Vestigial - for compat with Linux list code; rename to listp?
  */
 #define list_entry(LISTP, TYPE, FIELD) (LISTP)
 
 #define listp_for_each_entry(CURSOR, HEAD, FIELD)                       \
-    for(int first_iter = ({ (CURSOR) = (HEAD)->first;                   \
-                (HEAD)->first ? 1 : 0; });                              \
-        first_iter || (CURSOR) != (HEAD)->first;                        \
-        (CURSOR) = (CURSOR)->FIELD.next, first_iter = 0)
-
-#define listp_for_each_entry_reverse(CURSOR, HEAD, FIELD)              \
-    for(int first_iter = ({(CURSOR) = ((HEAD)->first                   \
-                    ? (HEAD)->first->FIELD.prev :                      \
-                    (HEAD)->first); (HEAD)->first ? 1 : 0; });         \
-        first_iter || ((CURSOR) && (CURSOR)->FIELD.next != (HEAD)->first); \
-        (CURSOR) = (CURSOR)->FIELD.prev, first_iter = 0)
-
-#define listp_for_each_entry_safe(CURSOR, TMP, HEAD, FIELD)             \
-    for(int first_iter = ({(CURSOR) = (HEAD)->first;                    \
-                    (TMP) = ((CURSOR) ? (CURSOR)->FIELD.next : (CURSOR)); \
-                (HEAD)->first ? 1 : 0; });                              \
-        (first_iter || (CURSOR) != (HEAD)->first) && (HEAD)->first;     \
-        first_iter = (first_iter && (TMP) != (CURSOR) && (HEAD)->first == (TMP) ? \
-                      1: 0),                                            \
-            (CURSOR) = (TMP), (TMP) = (TMP)->FIELD.next)
+    for (bool first_iter = ((CURSOR) = (HEAD)->first,                   \
```

Comment:
Is this a significant change to the list code, or just indentation or idiom swapping?

There are some unit tests in the lib directory that are not part of CI.  I would recommend running them, and testing the bah-jeebus out of any changes - this is highly error prone code, and I would recommend leaving it alone unless there is a serious reason to change it, or maybe even pull this into a separate PR.

Reply:
I think @mkow 's intention is just to first_iter a boolean. Otherwise, these two code snippets are interchangeable.

Reply:
Yup, @chiache is right. The old code was very hard to read and understand, this version should be simpler.

Reply:
I'm ok with legibility improvements, but you may still want to run the list unit tests before the merge just to be sure.

Reply:
Yup, I totally agree that we need to run unit tests before merging. Where are they located and how to run them?
Moving this change to a separate PR seems to be a good idea IMO.

Reply:
Pal/lib/list-test.c

Reply:
They seem to pass. Console log:
```
user@host:[vma-rewrite]~/graphene/Pal/lib$ gcc -std=gnu99 list-test.c -o list-test
user@host:[vma-rewrite]~/graphene/Pal/lib$ ./list-test 
Beginning of list
End of list

Deletion test starting

Deletion test Ending

Deletion test 2 starting

Deletion test 2 Ending

Begin splice tests 

Before list move test 

After list move test 

All tests passed!
user@host:[vma-rewrite]~/graphene/Pal/lib$ echo $?
0
```
Running them takes ~0 seconds, so maybe it'd be a good idea to enable them on Jenkins?